dali/python/nvidia/dali/external_source.py

```diff
@@ -0,0 +1,258 @@
+        if len(data) > 0:
+            dim = len(data[0].shape)
+            for t in data:
+                if len(t.shape) != dim:
+                    raise RuntimeErrorError("All tensors in a batch must have the same number of dimensions")
+            if layout != "" and dim != len(layout):
+                raise RuntimeError("The layout '{}' cannot describe {}-dimensional data".format(layout, dim))
+    else:
+        dim = len(data.shape) - 1
+        if data.shape[0] != batch_size:
+            raise RuntimeError("The external source returned an unexpected batch "
+            "size: {} instead of {}".format(data.shape[0], batch_size))
+        if layout != "" and dim != len(layout):
+            raise RuntimeError("The layout '{}' cannot describe {}-dimensional data".format(layout, dim))
+
+
+class _ExternalSourceGroup(object):
+    def __init__(self, callback, is_multioutput, instances = []):
+        self.instances = list(instances)  # we need a copy!
+        self.is_multioutput = is_multioutput
+        self.callback = callback
+        if callback is not None:
+            if callback.__code__.co_argcount not in [0, 1]:
+                raise TypeError("External source callback must be a callable with 0 or 1 argument")
+            self.accepts_iter_num = (callback.__code__.co_argcount == 1)
+        else:
+            self.accepts_iter_num = None
+
+    def append(self, instance):
+        self.instances.append(instance)
+
+    def call_and_feed(self, pipeline, current_iter):
+        if self.accepts_iter_num:
+            callback_out = self.callback(current_iter)
+        else:
+            callback_out = self.callback()
+
+        if self.is_multioutput:
+            for op in self.instances:
+                data = callback_out[op._output_index]
+                pipeline.feed_input(op._name, data, op._layout)
+        else:
+            data = callback_out
+            op = self.instances[0]
+            pipeline.feed_input(op._name, data, op._layout)
+
+def _get_callback_from_source(source, cycle):
+    iterable = False
+    if source is not None:
+        try:
+            iterator = iter(source)
+            if cycle:
+                iterator = itertools.cycle(iterator)
+            iterable = True
+            callback = lambda: next(iterator)
+        except TypeError:
+            if not callable(source):
+                raise TypeError("Source must be iterable or callable")
+            callback = source
+    else:
+        callback = None
+
+    if not iterable and cycle:
+        raise ValueError("`cycle` argument is only valid for iterable `source`")
+    return callback
+
+class ExternalSource(with_metaclass(_DaliOperatorMeta, object)):
+    """ExternalSource is a special operator which can provide data to DALI pipeline from Python
+    using several methods.
+
+    The simplest and preferred way is to specify a `source`, which may be a callable or iterable.
+
+    `source` : callable or iterable
```

Comment:
I'm not sure if you need to duplicate the argument comments twice, but probably you would need to use the
```
Args
------

Keyword Args
------------
```
sections.